go/sqlescape/ids.go

```diff
@@ -14,22 +14,33 @@ limitations under the License.
 package sqlescape
 
 import (
-	"bytes"
+	"strings"
 )
 
 // EscapeID returns a backticked identifier given an input string.
 func EscapeID(in string) string {
-	var buf bytes.Buffer
+	var buf strings.Builder
```

Comment:
I wonder if it wouldn’t be more efficient to just set the initial capacity to len(in)+4 here. That would cover the outer backticks and escaping of one inner. That should cover ~99.999999999999% of cases and we could skip the byte iteration and compare to count the backticks in the ident (I’ve never once seen a backtick used in a table name).

Reply:
I can try that and benchmark it. I don't think the first loop to count adds anything really measurable to the result relative to the memory allocation. Typically in Go for high performance things, it's still favored to do extra loops to allocate once correctly. So the iteration _should_ be cheap, but I'll try it out for sure. You're likely right though. It's likely cheaper to allocate slightly more and cover most cases just fine. I assume as long as it's within the same slab class (or whatever Go's allocater pages are called), it'd be the same efficiency.

Reply:
Ok, so interesting results.

before:

```
$ go test -v vitess.io/vitess/go/sqlescape -bench=. -benchmem
=== RUN   TestEscapeID
--- PASS: TestEscapeID (0.00s)
goos: linux
goarch: amd64
pkg: vitess.io/vitess/go/sqlescape
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkEscapeID
BenchmarkEscapeID/aa
BenchmarkEscapeID/aa-8          46028002                35.27 ns/op            4 B/op        1 allocs/op
BenchmarkEscapeID/a`a
BenchmarkEscapeID/a`a-8         20073259                53.98 ns/op            8 B/op        1 allocs/op
BenchmarkEscapeID/long
BenchmarkEscapeID/long-8         1790821               715.6 ns/op           144 B/op        1 allocs/op
PASS
```

after:

```
=== RUN   TestEscapeID
--- PASS: TestEscapeID (0.00s)
goos: linux
goarch: amd64
pkg: vitess.io/vitess/go/sqlescape
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkEscapeID
BenchmarkEscapeID/aa
BenchmarkEscapeID/aa-8          45377708                45.72 ns/op            8 B/op        1 allocs/op
BenchmarkEscapeID/a`a
BenchmarkEscapeID/a`a-8         24866187                48.87 ns/op            8 B/op        1 allocs/op
BenchmarkEscapeID/long
BenchmarkEscapeID/long-8         2398213               516.4 ns/op           160 B/op        1 allocs/op
PASS
```

With the diff: 

```diff
diff --git a/go/sqlescape/ids.go b/go/sqlescape/ids.go
index 9887ed2ccd..96ada5d2f2 100644
--- a/go/sqlescape/ids.go
+++ b/go/sqlescape/ids.go
@@ -26,19 +26,10 @@ func EscapeID(in string) string {
 
 // WriteEscapeID writes a backticked identifier from an input string into buf.
 func WriteEscapeID(buf *strings.Builder, in string) {
-	i := 0
-
-	// calculate the total length required after escaping
-	ll := 2 + len(in)
-	for i = 0; i < len(in); i++ {
-		if in[i] == '`' {
-			ll++
-		}
-	}
-	buf.Grow(ll)
+	buf.Grow(4 + len(in))
 
 	buf.WriteByte('`')
-	for i = 0; i < len(in); i++ {
+	for i := 0; i < len(in); i++ {
 		buf.WriteByte(in[i])
 		if in[i] == '`' {
 			buf.WriteByte('`')
```

So it's an objective improvement over _long_ identifiers because of obviously the extra O(N) loop.

We get still 1 allocation for a small identifier, but twice as large of an allocation (4B vs 8B) which makes sense since we're rounding up, which makes it decently slower, guessing because of the larger allocation.

So I guess we're at an improvement for long identifiers, but a loss for small identifiers, and likely a wash for some definition of average length.

Either way, I don't have a strong preference here, and I don't think it's worth adding in different logic to say `if len(in) > 20` or something and handle it differently.

But for posterity:

```
=== RUN   TestEscapeID
--- PASS: TestEscapeID (0.00s)
goos: linux
goarch: amd64
pkg: vitess.io/vitess/go/sqlescape
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkEscapeID
BenchmarkEscapeID/aa
BenchmarkEscapeID/aa-8          48326787                36.16 ns/op            4 B/op        1 allocs/op
BenchmarkEscapeID/a`a
BenchmarkEscapeID/a`a-8         36968911                53.68 ns/op            8 B/op        1 allocs/op
BenchmarkEscapeID/long
BenchmarkEscapeID/long-8         2654198               452.6 ns/op           160 B/op        1 allocs/op
PASS
```

Gets us kinda the best of both worlds with something like:

```diff
diff --git a/go/sqlescape/ids.go b/go/sqlescape/ids.go
index 9887ed2ccd..508d9daad2 100644
--- a/go/sqlescape/ids.go
+++ b/go/sqlescape/ids.go
@@ -26,19 +26,22 @@ func EscapeID(in string) string {
 
 // WriteEscapeID writes a backticked identifier from an input string into buf.
 func WriteEscapeID(buf *strings.Builder, in string) {
-	i := 0
-
-	// calculate the total length required after escaping
 	ll := 2 + len(in)
-	for i = 0; i < len(in); i++ {
-		if in[i] == '`' {
-			ll++
+
+	if len(in) < 20 {
+		for i := 0; i < len(in); i++ {
+			if in[i] == '`' {
+				ll++
+			}
 		}
+	} else {
+		ll += 2
 	}
+
 	buf.Grow(ll)
 
 	buf.WriteByte('`')
-	for i = 0; i < len(in); i++ {
+	for i := 0; i < len(in); i++ {
 		buf.WriteByte(in[i])
 		if in[i] == '`' {
 			buf.WriteByte('`')
```

I'm not sure the extra complexity for such a small gain is worth it though.